packages/core/src/electron-main/electron-main-application.ts

```diff
@@ -0,0 +1,319 @@
+/********************************************************************************
+ * Copyright (C) 2019 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import * as electron from 'electron';
+import nativeKeymap = require('native-keymap');
+import ElectronStorage = require('electron-store');
+import { injectable, inject, named } from 'inversify';
+import { ContributionProvider, MaybePromise } from '../common';
+import { Deferred } from '../common/promise-util';
+import { fork } from 'child_process';
+
+export const ElectronMainApplicationContribution = Symbol('ElectronMainApplicationContribution');
+export interface ElectronMainApplicationContribution {
+
+    onStart?(app: Electron.App): MaybePromise<void>;
+
+    /**
+     * Will wait for both the electron `ready` event, and all contributions to
+     * finish executing `onStart`.
+     *
+     * @param infos https://electronjs.org/docs/api/app#event-ready.
+     */
+    // tslint:disable-next-line:no-any
+    onReady?(infos?: any): MaybePromise<void>;
+
+    onNewWindow?(newWindow: electron.BrowserWindow): MaybePromise<void>;
+
+    onBeforeQuit?(event: electron.Event): MaybePromise<void>;
+
+    onQuit?(event: electron.Event, exitCode: number): MaybePromise<void>;
```

Comment:
Do you have any example with asynchronous `onQuit`? Does it work?

Reply:
Will make it sync for now indeed.